---   
---   

<h1 align="center">ExD</h1>
<h1 align="center">Course: Advanced Python Programming Language</h1>


---
<h3><div align="right">Instructor: Kiran Khursheed</div></h3>    

<h1 align="center">Pandas-04</h1>

## _IO with CSV EXCEL and JSON Files_

<img align="center" width="600" height="150"  src="images/fileformats.png" >

#### Read Pandas Documentation:
- General Info: https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html


- For `read_csv`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html?highlight=read_csv#pandas.read_csv


- For `read_excel`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html?highlight=read_excel#pandas.read_excel


- For `read_json`:https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.io.json.read_json.html?highlight=pandas%20read_json#pandas.io.json.read_json


- For `to_csv`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html#pandas.DataFrame.to_csv



- For `to_excel`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_excel.html?highlight=to_excel#pandas.DataFrame.to_excel


- For `to_json`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_json.html?highlight=to_json

## Learning agenda of this notebook
[Pandas](https://pandas.pydata.org/) provides helper functions to read data from various file formats like CSV, EXCEL, JSON, HTML, SQL table, and many more.
1. Reading data from a CSV/TSV File
2. Reading a CSV file from a Remote System
3. Writing Contents of Dataframe to a CSV File
4. Reading data from an EXCEL File
5. Writing Contents of Dataframe to an EXCEL File
6. Reading data from a JSON File
7. Writing Contents of Dataframe to a JSON File

In [ ]:
# To install this library in Jupyter notebook
import sys
!{sys.executable} -m pip install pandas --quiet

In [1]:
import pandas as pd
pd.__version__ , pd.__path__

('1.3.4',
 ['/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pandas'])

## 1. Reading from CSV/TSV Files
>**CSV**: A text file in which the values are separated by a comma or a tab character is called a CSV or a TSV file. Each line of the file is a data record and each record consists of one or more fields, separated by a specific character called separator. A CSV/TSV file is typically used to store tabular data (numbers and text), in which each line will have the same number of fields.

### a. Reading a  Simple CSV File
The `pd.read_csv()` method is used to read a comma-separated file into a DataFrame.
```
pd.read_csv(fname, delimiter=None, header='infer', skiprows=None , nrows=None , usecols=None,  footer='',...)

```

In [2]:
! cat datasets/classmarks.csv

rollno,gender,group,age,math,english,urdu
MS01,female,group B,28.0,72.0,72,74.0
MS02,female,group C,33.0,69.0,90,88.0
MS03,female,group B,21.0,,95,93.0
MS04,male,group A,44.0,47.0,57,44.0
MS05,male,group C,54.0,76.0,78,
MS06,female,group B,,71.0,83,78.0
MS07,female,group B,47.0,88.0,95,92.0
MS08,male,group B,33.0,40.0,43,39.0
MS09,male,group D,27.0,64.0,64,67.0
MS10,female,group B,33.0,38.0,60,50.0
MS11,male,group C,46.0,58.0,54,52.0
MS12,male,group D,53.0,40.0,52,43.0
MS13,female,group B,25.0,65.0,81,73.0
MS14,male,group A,26.0,78.0,72,70.0
MS15,female,group A,25.0,50.0,53,58.0
MS16,female,group C,29.0,69.0,75,78.0
MS17,male,group C,22.0,88.0,89,86.0
MS18,female,group B,31.0,18.0,32,28.0
MS19,male,group C,33.0,46.0,42,46.0
MS20,female,group C,36.0,54.0,58,
MS21,male,group D,34.0,66.0,69,63.0
MS22,female,group B,32.0,65.0,75,70.0
MS23,male,group D,31.0,44.0,54,53.0
MS24,female,group C,37.0,,73,73.0
MS25,male,group D,33.0,74.0,71,80.0
MS26,male,group A,39.0,73.0,74,72.0
MS27,male,group 

In [4]:
df.shape

(50, 7)

In [3]:
#The `read_csv`, by default assumes that the file contains comma separated values, 
# and the first row of the file conatins names of columns, which will be taken as column labels
df = pd.read_csv('datasets/classmarks.csv')
df

rollno  gender    group   age  math  english  urdu
0    MS01  female  group B  28.0  72.0       72  74.0
1    MS02  female  group C  33.0  69.0       90  88.0
2    MS03  female  group B  21.0   NaN       95  93.0
3    MS04    male  group A  44.0  47.0       57  44.0
4    MS05    male  group C  54.0  76.0       78   NaN
5    MS06  female  group B   NaN  71.0       83  78.0
6    MS07  female  group B  47.0  88.0       95  92.0
7    MS08    male  group B  33.0  40.0       43  39.0
8    MS09    male  group D  27.0  64.0       64  67.0
9    MS10  female  group B  33.0  38.0       60  50.0
10   MS11    male  group C  46.0  58.0       54  52.0
11   MS12    male  group D  53.0  40.0       52  43.0
12   MS13  female  group B  25.0  65.0       81  73.0
13   MS14    male  group A  26.0  78.0       72  70.0
14   MS15  female  group A  25.0  50.0       53  58.0
15   MS16  female  group C  29.0  69.0       75  78.0
16   MS17    male  group C  22.0  88.0       89  86.0
17   MS18  female  group B  31.0  18.0       32  28.0
18   MS19    male  group C  33.0  46.0       42  46.0
19   MS20  female  group C  36.0  54.0       58   NaN
20   MS21    male  group D  34.0  66.0       69  63.0
21   MS22  female  group B  32.0  65.0       75  70.0
22   MS23    male  group D  31.0  44.0       54  53.0
23   MS24  female  group C  37.0   NaN       73  73.0
24   MS25    male  group D  33.0  74.0       71  80.0
25   MS26    male  group A  39.0  73.0       74  72.0
26   MS27    male  group B   NaN  69.0       54  55.0
27   MS28  female  group C  47.0  67.0       69  75.0
28   MS29    male  group C  44.0  70.0       70  65.0
29   MS30  female  group D  42.0  62.0       70  75.0
30   MS31  female  group D  41.0  69.0       74  74.0
31   MS32  female  group B   NaN  63.0       65  61.0
32   MS33  female  group E  29.0  56.0       72  65.0
33   MS34    male  group D  34.0  40.0       42  38.0
34   MS35    male  group E  45.0  97.0       87  82.0
35   MS36    male  group E  31.0  81.0       81  79.0
36   MS37  female  group D  28.0  74.0       81   NaN
37   MS38  female  group D  26.0  50.0       64  59.0
38   MS39  female  group D  36.0  75.0       90  88.0
39   MS40    male  group B  33.0  57.0       56  57.0
40   MS41    male  group C  35.0  55.0       61  54.0
41   MS42  female  group C  22.0  58.0       73  68.0
42   MS43  female  group B  38.0  53.0       58  65.0
43   MS44    male  group B  31.0  59.0       65  66.0
44   MS45  female  group E  28.0   NaN       56  54.0
45   MS46    male  group B  26.0  65.0       54  57.0
46   MS47  female  group A  36.0  55.0       65  62.0
47   MS48  female  group C  30.0  66.0       71  76.0
48   MS49  female  group D  40.0  57.0       74  76.0
49   MS50    male  group C  37.0  66.0       78  81.0

**The `df.head(N)` method is used to select/display first `N` rows, based on `position`, i.e., the integer value corresponding to the position of the row (from 0 to n-1). The default value of `N` is 5.**

In [ ]:
df.head()

In [ ]:
df.head(3)

In [ ]:
# For negative values of n, this method returns all rows except the last `n` rows, equivalent to df[:-n].
# The df has a total of 50 rows, so the following will return first 2 rows
df.head(-48)

**The `df.tail(N)` method is used to select/display last `N` rows, based on `position`, i.e., the integer value corresponding to the position of the row (from 0 to n-1). The default value of `N` is 5.**

In [ ]:
# tail() method is useful for quickly verifying data, after sorting or appending rows.
df.tail()

In [ ]:
df.tail(3)

In [ ]:
# For negative values of `n`, this function returns all rows except the first `n` rows, equivalent to df[n:]
# The df has a total of 50 rows, so the following will return last 3 rows
df.tail(-46)

### b.Reading a CSV File having a Delimter, other than Comma
- By default, the `read_csv()` expect comma as seperator. But if the CSV file has some other seperator or delimiter like (semi-collon or tab), it will raise an error.
- To handler the issue we need to pass specific value to the `delimiter` argument of `read_csv()` method.

In [ ]:
! cat datasets/classmarkswithtab.csv

In [5]:
df = pd.read_csv('datasets/classmarkswithtab.csv')
df.head()

rollno\tgender\tgroup\tage\tmath\tenglish\turdu
0     MS01\tfemale\tgroup B\t28.0\t72.0\t72\t74.0
1     MS02\tfemale\tgroup C\t33.0\t69.0\t90\t88.0
2         MS03\tfemale\tgroup B\t21.0\t\t95\t93.0
3       MS04\tmale\tgroup A\t44.0\t47.0\t57\t44.0
4           MS05\tmale\tgroup C\t54.0\t76.0\t78\t

In [13]:
df = pd.read_csv('datasets/classmarkswithtab.csv', delimiter='\t', header=None)
df.head()

0       1        2     3     4        5     6
0  rollno  gender    group   age  math  english  urdu
1    MS01  female  group B  28.0  72.0       72  74.0
2    MS02  female  group C  33.0  69.0       90  88.0
3    MS03  female  group B  21.0   NaN       95  93.0
4    MS04    male  group A  44.0  47.0       57  44.0

### c. Reading a CSV File not having Column Labels
- By default the `read_csv()` method assume the first row of the file will contain column labels
- If this is not the case, i.e., the file do not contain column labels rather data, it will be dealt as column label
- Understand this in following example

In [ ]:
! cat datasets/classmarkswithoutcollabels.csv

In [7]:
df = pd.read_csv('datasets/classmarkswithoutcollabels.csv')
df.head()

MS01  female  group B    28    72  72.1    74
0  MS02  female  group C  33.0  69.0    90  88.0
1  MS03  female  group B  21.0   NaN    95  93.0
2  MS04    male  group A  44.0  47.0    57  44.0
3  MS05    male  group C  54.0  76.0    78   NaN
4  MS06  female  group B   NaN  71.0    83  78.0

**To read such files, you have to pass the parameter `header=None` to the `read_csv()` method as shown below**

In [8]:
df = pd.read_csv('datasets/classmarkswithoutcollabels.csv', header=None)
df.head()


0       1        2     3     4   5     6
0  MS01  female  group B  28.0  72.0  72  74.0
1  MS02  female  group C  33.0  69.0  90  88.0
2  MS03  female  group B  21.0   NaN  95  93.0
3  MS04    male  group A  44.0  47.0  57  44.0
4  MS05    male  group C  54.0  76.0  78   NaN

**Now if you want to assign new column labels to make them more understandable, you can assign the list of column labels to the `columns` attribute of the dataframe object**

In [9]:
col_names = ['rollno', 'gender', 'group', 'age', 'math', 'english', 'urdu']
df.columns = col_names
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

### d. Reading a CSV File having Comments in the beginning
- You may get an error while reading a CSV file because someone may have added few comments on the top of the file. In pandas we can still read the data set by skipping few rows from the top.
- To deal with the ParseError, open the csv file in the text editor and check if you have some comments on the top.
- If yes, then count the number of rows to skip.
- While reading file, pass the parameter **skiprows = n** (number of rows in the beginninghaving comments to skip)
- While reading file, pass the parameter **skipfooter = n** (number of rows at the end having comments to skip)

In [10]:
! cat datasets/classmarkswithtopcomments.csv






















































MS50,male,group C,37,66,78,81

In [11]:
# Try reading a csv file having 3 comments lines in the beginning.
df = pd.read_csv('datasets/classmarkswithtopcomments.csv')
df.head()

These are comments Unnamed: 1 Unnamed: 2 Unnamed: 3  \
0  These are second line of comments        NaN        NaN        NaN   
1   These are third line of comments        NaN        NaN        NaN   
2                             rollno     gender      group        age   
3                               MS01     female    group B         28   
4                               MS02     female    group C         33   

  Unnamed: 4 Unnamed: 5 Unnamed: 6  
0        NaN        NaN        NaN  
1        NaN        NaN        NaN  
2       math    english       urdu  
3         72         72         74  
4         69         90         88

In [3]:
# Try reading a csv file having 3 comments lines in the beginning.
import pandas as pd
df = pd.read_csv('datasets/classmarkswithtopcomments.csv', skiprows=3)
df.head(10)

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN
5   MS06  female  group B   NaN  71.0       83  78.0
6   MS07  female  group B  47.0  88.0       95  92.0
7   MS08    male  group B  33.0  40.0       43  39.0
8   MS09    male  group D  27.0  64.0       64  67.0
9   MS10  female  group B  33.0  38.0       60  50.0

### e. Reading a portion of CSV File in a Dataframe
- Suppose the dataset inside the csv file is too big and you don't want to spend that much time for reading that data
- Or might be your system crashes, when you try to load that much data
- Solution is read
    - Specific number of rows by passing `nrows` parameter to `read_csv()` method
    - Specific number of columns by passing `usecols` parameter to `read_csv()` method

In [12]:
# Read just 10 rows from the csv file by passing the number of rows to read to `nrows` argument
df = pd.read_csv('datasets/classmarks.csv', nrows=10)
df.shape
#df.head()

(10, 7)

In [5]:
# Read specific columns from the csv file by passing a list of column names to `usecols` argument
df = pd.read_csv('datasets/classmarks.csv', usecols= ['rollno', 'group','english'])
df.shape
df.head()

rollno    group  english
0   MS01  group B       72
1   MS02  group C       90
2   MS03  group B       95
3   MS04  group A       57
4   MS05  group C       78

In [14]:
df.index 

RangeIndex(start=0, stop=50, step=1)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   rollno   50 non-null     object
 1   group    50 non-null     object
 2   english  50 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 1.3+ KB


In [8]:
df.dtypes

rollno     object
group      object
english     int64
dtype: object

In [9]:
df.empty

False

In [10]:
df.columns

Index(['rollno', 'group', 'english'], dtype='object')

In [11]:
df.size

150

In [12]:
df.ndim

2

In [13]:
df.shape

(50, 3)

In [13]:
# Ofcourse you can use both the parameters at the same time
df = pd.read_csv('datasets/classmarks.csv', 
                 nrows= 7, usecols= ['rollno', 'group','english'])
df.shape

(7, 3)

In [14]:
df

rollno    group  english
0   MS01  group B       72
1   MS02  group C       90
2   MS03  group B       95
3   MS04  group A       57
4   MS05  group C       78
5   MS06  group B       83
6   MS07  group B       95

## 2. Reading a CSV File from a Remote System

In [15]:
# To avoid URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED]..... 
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

### a. Reading a CSV file from GitHub Gist

The `myfamily.csv` file actually resides on my GitHub Gist at following URL:
https://gist.githubusercontent.com/kiranzaman/79bac8e2ca947c7db5f25598ce85fc9d/raw/bf7c14aeaa68d644fda770896928bfa9b9b159e9/myfamily.csv

Bitly is a URL shortening service that I have used to create a short link for easy usage in the cell below:

In [1]:
import pandas as pd
myurl = 'https://gist.githubusercontent.com/kiranzaman/79bac8e2ca947c7db5f25598ce85fc9d/raw/bf7c14aeaa68d644fda770896928bfa9b9b159e9/myfamily.csv'
df = pd.read_csv(myurl)
df

Names  Age       Addr
0  Khursheed   75     Lahore
1       Aqsa   29  Islamabad
2       Sana   28   Peshawer
3    Wajeeha   27  Islamabad
4    Basirat   20     Lahore
5        Hrm   18    Karachi

### b. Reading a CSV file from a Google Docs

- Google Sheet URL:  https://docs.google.com/spreadsheets/d/1q_xkjkoz5_t_HhrhZWEr__bj7UzpDeVQ-kT5EUe4Ay0/edit?gid=0#gid=0


In [4]:
sheetID = '1q_xkjkoz5_t_HhrhZWEr__bj7UzpDeVQ-kT5EUe4Ay0'
sheetName = 'Sheet1'
URL = 'https://docs.google.com/spreadsheets/d/{0}/gviz/tq?tqx=out:csv&sheet={1}'.format(sheetID, sheetName)

df = pd.read_csv(URL)
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

In [18]:
URL

'https://docs.google.com/spreadsheets/d/1H9ZTGVRXN3zuyP3cbJQnbX7JlqhqFNMOE_WwDP-PRTE/gviz/tq?tqx=out:csv&sheet=sheet1'

## 3. Writing Contents of Dataframe to a CSV File
- The `pd.to_csv()` method is used to write the contents of a dataframe (with indices) to a CSV file.
- The only required argument is the file path.
- For details see help page or python documents (link given above)

In [15]:
df_class = pd.read_csv('datasets/classmarks.csv')
df_class.head(7)


rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN
5   MS06  female  group B   NaN  71.0       83  78.0
6   MS07  female  group B  47.0  88.0       95  92.0

>- Let us create a new dataframe from above dataframe containing records of only group B

In [16]:
mask = (df_class['group'] == 'group B')
mask.head(7)

0     True
1    False
2     True
3    False
4    False
5     True
6     True
Name: group, dtype: bool

In [18]:
df_class_groupB = df_class.loc[df_class['group'] == 'group B']
df_class_groupB

rollno  gender    group   age  math  english  urdu
0    MS01  female  group B  28.0  72.0       72  74.0
2    MS03  female  group B  21.0   NaN       95  93.0
5    MS06  female  group B   NaN  71.0       83  78.0
6    MS07  female  group B  47.0  88.0       95  92.0
7    MS08    male  group B  33.0  40.0       43  39.0
9    MS10  female  group B  33.0  38.0       60  50.0
12   MS13  female  group B  25.0  65.0       81  73.0
17   MS18  female  group B  31.0  18.0       32  28.0
21   MS22  female  group B  32.0  65.0       75  70.0
26   MS27    male  group B   NaN  69.0       54  55.0
31   MS32  female  group B   NaN  63.0       65  61.0
39   MS40    male  group B  33.0  57.0       56  57.0
42   MS43  female  group B  38.0  53.0       58  65.0
43   MS44    male  group B  31.0  59.0       65  66.0
45   MS46    male  group B  26.0  65.0       54  57.0

In [20]:
df_class_groupB.to_csv('datasets/classmarksgroupB.csv')

In [21]:
df = pd.read_csv('datasets/classmarksgroupB.csv')
df

Unnamed: 0 rollno  gender    group   age  math  english  urdu
0            0   MS01  female  group B  28.0  72.0       72  74.0
1            2   MS03  female  group B  21.0   NaN       95  93.0
2            5   MS06  female  group B   NaN  71.0       83  78.0
3            6   MS07  female  group B  47.0  88.0       95  92.0
4            7   MS08    male  group B  33.0  40.0       43  39.0
5            9   MS10  female  group B  33.0  38.0       60  50.0
6           12   MS13  female  group B  25.0  65.0       81  73.0
7           17   MS18  female  group B  31.0  18.0       32  28.0
8           21   MS22  female  group B  32.0  65.0       75  70.0
9           26   MS27    male  group B   NaN  69.0       54  55.0
10          31   MS32  female  group B   NaN  63.0       65  61.0
11          39   MS40    male  group B  33.0  57.0       56  57.0
12          42   MS43  female  group B  38.0  53.0       58  65.0
13          43   MS44    male  group B  31.0  59.0       65  66.0
14          45   MS46    male  group B  26.0  65.0       54  57.0

>To avoid writing the row indices column inside the file pass `index=False` argument to `to_csv()` method

In [ ]:
df_class_groupB.to_csv('datasets/classmarksgroupB.csv')
print(df_class_groupB)
df_class_groupB.to_csv('datasets/classmarksgroupB.csv', index=False)
print(df_class_groupB)

In [26]:
%pwd

'C:\\Users\\Kiran\\AdvancedPython\\Pandas'

In [24]:
df_class_groupB.to_csv('classmarksgroupB.csv', index=False)

In [25]:
df = pd.read_csv('datasets/classmarksgroupB.csv')
df

rollno  gender    group   age  math  english  urdu
0    MS01  female  group B  28.0  72.0       72  74.0
1    MS03  female  group B  21.0   NaN       95  93.0
2    MS06  female  group B   NaN  71.0       83  78.0
3    MS07  female  group B  47.0  88.0       95  92.0
4    MS08    male  group B  33.0  40.0       43  39.0
5    MS10  female  group B  33.0  38.0       60  50.0
6    MS13  female  group B  25.0  65.0       81  73.0
7    MS18  female  group B  31.0  18.0       32  28.0
8    MS22  female  group B  32.0  65.0       75  70.0
9    MS27    male  group B   NaN  69.0       54  55.0
10   MS32  female  group B   NaN  63.0       65  61.0
11   MS40    male  group B  33.0  57.0       56  57.0
12   MS43  female  group B  38.0  53.0       58  65.0
13   MS44    male  group B  31.0  59.0       65  66.0
14   MS46    male  group B  26.0  65.0       54  57.0

In [30]:
df = pd.read_csv('datasets/groupdatawithoutcollables.csv', header = None)
df.head(3)

0        1   2          3          4        5       6     7     8  \
0  MS01     Rauf  52     Lahore    MORNING  group C    Male  78.3  84.4   
1  MS02     Arif  51  Islamabad        AFT  group A    Male  70.5  60.5   
2  MS03  Shaista  35    Karachi  AFTERNOON  group B  Female  64.9  75.1   

        9  
0  5000.0  
1  6000.0  
2  8500.0

In [31]:
df.columns = ['roll no', 'name', 'age', 'address', 'session', 'group', 'gender', 'subj1', 'subj2', 'scholarship']
df.head(3)

roll no     name  age    address    session    group  gender  subj1  subj2  \
0    MS01     Rauf   52     Lahore    MORNING  group C    Male   78.3   84.4   
1    MS02     Arif   51  Islamabad        AFT  group A    Male   70.5   60.5   
2    MS03  Shaista   35    Karachi  AFTERNOON  group B  Female   64.9   75.1   

   scholarship  
0       5000.0  
1       6000.0  
2       8500.0

In [32]:
df.columns.str.replace(' ', '_')

Index(['roll_no', 'name', 'age', 'address', 'session', 'group', 'gender',
       'subj1', 'subj2', 'scholarship'],
      dtype='object')

In [34]:
df.columns = df.columns.str.replace(' ', '_')

In [35]:
df = pd.read_csv('datasets/groupdata.csv')
df.head(3)

roll no     name  age    address    session    group  gender  subj1  subj2  \
0    MS01     Rauf   52     Lahore    MORNING  group C    Male   78.3   84.4   
1    MS02     Arif   51  Islamabad        AFT  group A    Male   70.5   60.5   
2    MS03  Shaista   35    Karachi  AFTERNOON  group B  Female   64.9   75.1   

   scholarship  
0       5000.0  
1       6000.0  
2       8500.0

In [37]:
#Since the inplace argument is by default False, so the rename() method will return a new dataframe
df.rename(mapper={'roll no': 'rollno', 'name':'fname'}, axis=1) #, inplace=True)

In [38]:
df.columns

Index(['rollno', 'fname', 'age', 'address', 'session', 'group', 'gender',
       'subj1', 'subj2', 'scholarship'],
      dtype='object')

In [39]:
import pandas as pd
df = pd.read_csv('datasets/groupdata.csv')
df.head(3)

roll no     name  age    address    session    group  gender  subj1  subj2  \
0    MS01     Rauf   52     Lahore    MORNING  group C    Male   78.3   84.4   
1    MS02     Arif   51  Islamabad        AFT  group A    Male   70.5   60.5   
2    MS03  Shaista   35    Karachi  AFTERNOON  group B  Female   64.9   75.1   

   scholarship  
0       5000.0  
1       6000.0  
2       8500.0

In [40]:
df['name'].map(len)

0     4
1     4
2     7
3     6
4     4
5     5
6     5
7     6
8     5
9     6
10    7
11    4
12    7
13    4
14    6
15    9
Name: name, dtype: int64

In [41]:
df.name.map(len)

0     4
1     4
2     7
3     6
4     4
5     5
6     5
7     6
8     5
9     6
10    7
11    4
12    7
13    4
14    6
15    9
Name: name, dtype: int64

In [42]:
df['len_name'] = df.name.map(len)

In [43]:
df.head()

roll no     name  age    address    session    group  gender  subj1  subj2  \
0    MS01     Rauf   52     Lahore    MORNING  group C    Male   78.3   84.4   
1    MS02     Arif   51  Islamabad        AFT  group A    Male   70.5   60.5   
2    MS03  Shaista   35    Karachi  AFTERNOON  group B  Female   64.9   75.1   
3    MS04   Hadeed   20     Lahore        MOR  group A    Male   82.0   84.3   
4    MS05     Zara   40   Peshawer        AFT  group D  Female   65.9   72.8   

   scholarship  len_name  
0       5000.0         4  
1       6000.0         4  
2       8500.0         7  
3       4000.0         6  
4       3500.0         4

In [44]:
df['session'].map({'MORNING':'M', 'AFTERNOON':'A'})

0       M
1     NaN
2       A
3     NaN
4     NaN
5       M
6     NaN
7       M
8     NaN
9       A
10    NaN
11      A
12      M
13      A
14    NaN
15      A
Name: session, dtype: object

In [45]:
df['session'].replace({'MORNING':'M', 'AFTERNOON':'A'})

0       M
1     AFT
2       A
3     MOR
4     AFT
5       M
6     AFT
7       M
8     AFT
9       A
10    MOR
11      A
12      M
13      A
14    AFT
15      A
Name: session, dtype: object

In [46]:
df['name'].apply(len)

0     4
1     4
2     7
3     6
4     4
5     5
6     5
7     6
8     5
9     6
10    7
11    4
12    7
13    4
14    6
15    9
Name: name, dtype: int64

In [47]:
# Let us use Lambda function to convert each name under the name column of df to upper case
df['name'].apply(lambda x : x.upper())

0          RAUF
1          ARIF
2       SHAISTA
3        HADEED
4          ZARA
5         MOHID
6         ZOBIA
7        IDREES
8         JAMIL
9        SHAHID
10      KHURRAM
11         MAAZ
12      MUJAHID
13         SARA
14       FATIMA
15    KAKAMANNA
Name: name, dtype: object

In [48]:
import pandas as pd
df = pd.read_csv('datasets/groupdata.csv')
df.head()

roll no     name  age    address    session    group  gender  subj1  subj2  \
0    MS01     Rauf   52     Lahore    MORNING  group C    Male   78.3   84.4   
1    MS02     Arif   51  Islamabad        AFT  group A    Male   70.5   60.5   
2    MS03  Shaista   35    Karachi  AFTERNOON  group B  Female   64.9   75.1   
3    MS04   Hadeed   20     Lahore        MOR  group A    Male   82.0   84.3   
4    MS05     Zara   40   Peshawer        AFT  group D  Female   65.9   72.8   

   scholarship  
0       5000.0  
1       6000.0  
2       8500.0  
3       4000.0  
4       3500.0

In [49]:
df.columns

Index(['roll no', 'name', 'age', 'address', 'session', 'group', 'gender',
       'subj1', 'subj2', 'scholarship'],
      dtype='object')

In [50]:
df.subj1 + df.subj2

0     162.7
1     131.0
2     140.0
3     166.3
4     138.7
5     147.9
6       NaN
7     160.1
8     171.8
9     171.8
10    171.8
11    171.8
12      NaN
13    160.1
14    171.8
15    171.8
dtype: float64

In [51]:
df['total']  = df.subj1.add(df.subj2, fill_value=0)


In [52]:
df.head()

roll no     name  age    address    session    group  gender  subj1  subj2  \
0    MS01     Rauf   52     Lahore    MORNING  group C    Male   78.3   84.4   
1    MS02     Arif   51  Islamabad        AFT  group A    Male   70.5   60.5   
2    MS03  Shaista   35    Karachi  AFTERNOON  group B  Female   64.9   75.1   
3    MS04   Hadeed   20     Lahore        MOR  group A    Male   82.0   84.3   
4    MS05     Zara   40   Peshawer        AFT  group D  Female   65.9   72.8   

   scholarship  total  
0       5000.0  162.7  
1       6000.0  131.0  
2       8500.0  140.0  
3       4000.0  166.3  
4       3500.0  138.7

In [53]:
del df['total']

In [54]:
df.pop('address')

0        Lahore
1     Islamabad
2       Karachi
3        Lahore
4      Peshawer
5        Lahore
6       Sialkot
7        Multan
8       Karachi
9        Lahore
10    Islamabad
11      Karachi
12       Lahore
13       Multan
14      Sialkot
15       Multan
Name: address, dtype: object

In [55]:
df.head()

roll no     name  age    session    group  gender  subj1  subj2  scholarship
0    MS01     Rauf   52    MORNING  group C    Male   78.3   84.4       5000.0
1    MS02     Arif   51        AFT  group A    Male   70.5   60.5       6000.0
2    MS03  Shaista   35  AFTERNOON  group B  Female   64.9   75.1       8500.0
3    MS04   Hadeed   20        MOR  group A    Male   82.0   84.3       4000.0
4    MS05     Zara   40        AFT  group D  Female   65.9   72.8       3500.0

In [14]:
df.apply(lambda col : col.isnull().sum() >= 2 )

roll no        False
session        False
group          False
gender         False
subj1          False
subj2          False
scholarship     True
dtype: bool

In [56]:
import pandas as pd
df = pd.read_csv('datasets/groupdata.csv')
df.drop(columns=['age', 'address', 'name'], axis=1, inplace=True)
df.head()

roll no    session    group  gender  subj1  subj2  scholarship
0    MS01    MORNING  group C    Male   78.3   84.4       5000.0
1    MS02        AFT  group A    Male   70.5   60.5       6000.0
2    MS03  AFTERNOON  group B  Female   64.9   75.1       8500.0
3    MS04        MOR  group A    Male   82.0   84.3       4000.0
4    MS05        AFT  group D  Female   65.9   72.8       3500.0

In [57]:
df.drop(index=[2,4], axis=0, inplace = True)

In [58]:
df.head()

roll no  session    group  gender  subj1  subj2  scholarship
0    MS01  MORNING  group C    Male   78.3   84.4       5000.0
1    MS02      AFT  group A    Male   70.5   60.5       6000.0
3    MS04      MOR  group A    Male   82.0   84.3       4000.0
5    MS06  MORNING  group C  Female   69.3   78.6          NaN
6    MS07      AFT  group B  Female   90.2    NaN       4000.0

In [ ]:
# It will delete the Scholarship column
df.drop(columns=df.columns[df.apply(lambda col: col.isnull().sum() >= 2)],\
        axis=1, inplace=True)

In [59]:
df = pd.read_csv('datasets/groupdata.csv')
df.dtypes

roll no         object
name            object
age              int64
address         object
session         object
group           object
gender          object
subj1          float64
subj2          float64
scholarship    float64
dtype: object

In [60]:
#Suppose we want to change the datatype of `age` column to float64 dtype
df['age'] = df.age.astype(float)
df.dtypes

roll no         object
name            object
age            float64
address         object
session         object
group           object
gender          object
subj1          float64
subj2          float64
scholarship    float64
dtype: object

In [79]:
# import the pandas library
import pandas as pd
df = pd.read_csv('datasets/group-marks.csv')
df.head()

rollno     name  gender    group    session  age  scholarship     math  \
0   MS01   SAADIA  female  group B    MORNING   28         2562  No Idea   
1   MS02  JUMAIMA  female  group C  AFTERNOON   33         2800       69   
2   MS03    ARIFA  female      NaN    EVENING   34         3500      NaN   
3   MS04   SAADIA  female  group A        MOR   44         2000       47   
4   MS05   DANISH    male  group C  AFTERNOON   54         2100       76   

   english  urdu  
0     72.0    74  
1     90.0    88  
2     95.0    93  
3     57.0    44  
4     78.0    55

In [80]:
df.dtypes

rollno          object
name            object
gender          object
group           object
session         object
age              int64
scholarship      int64
math            object
english        float64
urdu             int64
dtype: object

In [81]:
import numpy as np 
df.replace('No Idea', np.nan, inplace=True)

In [82]:
df.dtypes

rollno          object
name            object
gender          object
group           object
session         object
age              int64
scholarship      int64
math            object
english        float64
urdu             int64
dtype: object

In [83]:
# It is still Object, which is natural, however, we can change the datatype to `df.astype()` method
df['math'] = df['math'].astype(float)

In [69]:
df.math.mean() 

61.644444444444446

In [70]:
# Let us replace these values with mean value of the math column
df.loc[(df.math.isna()),'math'] = df.math.mean()

In [71]:
df.head()

rollno     name  gender    group    session  age  scholarship       math  \
0   MS01   SAADIA  female  group B    MORNING   28         2562  61.644444   
1   MS02  JUMAIMA  female  group C  AFTERNOON   33         2800  69.000000   
2   MS03    ARIFA  female      NaN    EVENING   34         3500  61.644444   
3   MS04   SAADIA  female  group A        MOR   44         2000  47.000000   
4   MS05   DANISH    male  group C  AFTERNOON   54         2100  76.000000   

   english  urdu  
0     72.0    74  
1     90.0    88  
2     95.0    93  
3     57.0    44  
4     78.0    55

In [74]:
# Confirm the result
df.isna().sum()
#df.info()

rollno         0
name           0
gender         0
group          3
session        0
age            0
scholarship    0
math           0
english        3
urdu           0
dtype: int64

In [84]:
# Let us read the dataset again with NA values
import pandas as pd
df = pd.read_csv('datasets/group-marks.csv', na_values='No Idea')
df.head(20)

rollno      name  gender    group    session  age  scholarship  math  \
0    MS01    SAADIA  female  group B    MORNING   28         2562   NaN   
1    MS02   JUMAIMA  female  group C  AFTERNOON   33         2800  69.0   
2    MS03     ARIFA  female      NaN    EVENING   34         3500   NaN   
3    MS04    SAADIA  female  group A        MOR   44         2000  47.0   
4    MS05    DANISH    male  group C  AFTERNOON   54         2100  76.0   
5    MS06     SAFIA  female  group B        AFT   23         3800   NaN   
6    MS07      SARA  female  group B    EVENING   47         3000  88.0   
7    MS08  ABDULLAH    male  group B        EVE   33         2000  40.0   
8    MS09      KHAN    male  group D    MORNING   27         2500  64.0   
9    MS10   HASEENA  female  group B        AFT   33         2800  38.0   
10   MS11   MUSTJAB    male  group C        MOR   46         3000  58.0   
11   MS12     ABRAR    male  group D    MORNING   53         3312  40.0   
12   MS13    MAHOOR  female      NaN        MOR   25         2345  65.0   
13   MS14     USAMA    male  group A  AFTERNOON   26         2654  78.0   
14   MS15   NAVAIRA  female  group A        AFT   25         2137  50.0   
15   MS16   SAWAIRA  female  group C    EVENING   29         2567  69.0   
16   MS17     NOFIL    male  group C        MOR   22         3500  88.0   
17   MS18  SHUMAILA  female  group B  AFTERNOON   31         2500  18.0   
18   MS19  ABUBAKAR    male  group C        AFT   33         3000  46.0   
19   MS20     NADRA  female  group C    MORNING   36         2200  54.0   

    english  urdu  
0      72.0    74  
1      90.0    88  
2      95.0    93  
3      57.0    44  
4      78.0    55  
5      83.0    78  
6      95.0    92  
7      43.0    39  
8       NaN    67  
9      60.0    50  
10     54.0    52  
11     52.0    43  
12     81.0    73  
13     72.0    70  
14     53.0    58  
15     75.0    78  
16     89.0    86  
17      NaN    28  
18     42.0    46  
19     58.0    45

In [85]:
df.isna().sum()

rollno         0
name           0
gender         0
group          3
session        0
age            0
scholarship    0
math           5
english        3
urdu           0
dtype: int64

In [86]:
# forward fill or ffill attribute
# If have NaN value, just carry forward the previous value
# using ffill attribute, you can fill the NaN value with the previous value in that column
df.fillna(method = 'ffill', inplace=True)
df.head()

rollno     name  gender    group    session  age  scholarship  math  \
0   MS01   SAADIA  female  group B    MORNING   28         2562   NaN   
1   MS02  JUMAIMA  female  group C  AFTERNOON   33         2800  69.0   
2   MS03    ARIFA  female  group C    EVENING   34         3500  69.0   
3   MS04   SAADIA  female  group A        MOR   44         2000  47.0   
4   MS05   DANISH    male  group C  AFTERNOON   54         2100  76.0   

   english  urdu  
0     72.0    74  
1     90.0    88  
2     95.0    93  
3     57.0    44  
4     78.0    55

In [87]:
df.fillna(method = 'bfill', inplace=True)
df.head()

rollno     name  gender    group    session  age  scholarship  math  \
0   MS01   SAADIA  female  group B    MORNING   28         2562  69.0   
1   MS02  JUMAIMA  female  group C  AFTERNOON   33         2800  69.0   
2   MS03    ARIFA  female  group C    EVENING   34         3500  69.0   
3   MS04   SAADIA  female  group A        MOR   44         2000  47.0   
4   MS05   DANISH    male  group C  AFTERNOON   54         2100  76.0   

   english  urdu  
0     72.0    74  
1     90.0    88  
2     95.0    93  
3     57.0    44  
4     78.0    55

In [88]:
# Confirm the result
df.isna().sum()

rollno         0
name           0
gender         0
group          0
session        0
age            0
scholarship    0
math           0
english        0
urdu           0
dtype: int64

## 4. I/O with EXCEL Files
>**XLSX**: XLSX is a Microsoft Excel Open XML file format. It also comes under the Spreadsheet file format. It is an XML-based file format created by Microsoft Excel. In XLSX data is organized under the cells and columns in a sheet. Each XLSX file may contain one or more sheets. So a workbook can contain multiple sheets

In [ ]:
import sys
!{sys.executable} -m pip install xlrd xlwt openpyxl

### a. Reading a Simple Excel File

In [19]:
df = pd.read_excel(io='datasets/classmarks.xlsx')
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

### b. Reading an Excel File having Comments in the beginning
- You may get an error while reading an Excel file because someone may have added few comments on the top of the file. In pandas we can still read the data set by skipping few rows from the top.
- To deal with the ParseError, open the Excel file in MS EXCEL and check if you have some comments on the top.
- If yes, then count the number of rows to skip.
- While reading file, pass the parameter **skiprows = n** (number of rows in the beginning having comments to skip)
- While reading file, pass the parameter **skipfooter = n** (number of rows at the end having comments to skip)

In [ ]:
# The following file has three lines of comments in the beginning of the file.
df = pd.read_excel(io='datasets/classmarkswithcomments.xlsx',skiprows=3)

df.head()

### c. Reading Excel Workbook with Multiple Sheets
- By default `pd.read_excel()` function read only the first sheet.
- What if we want to read an Excel file having multiple sheets.
- The `big_mart_sales_with_multiple_sheets.xlsx` is a workbook that contains three sheets for different years data. The sheet names are 1985, 1987, and 1997

In [20]:
df = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx')
# if you check/view the data you can see, it only contains the data of first excel sheet (for the year 1985)
df.shape

(1463, 12)

In [21]:
df_1985 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx'
                        ,sheet_name='1985')
df_1987 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx',sheet_name='1987')
df_1997 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx',sheet_name='1997')

In [22]:
df_1985.shape

(1463, 12)

In [23]:
df_1987.shape

(932, 12)

In [24]:
df_1997.shape

(930, 12)

## 5. Writing Contents of Dataframe to an EXCEL File
- The `pd.to_excel()` method is used to write the contents of a dataframe (with indices) to an Excel file.
- The only required argument is the file path.
- For details see help page or python documents (link given above)

>- Let us create a new single dataframe after concatenating all the above three dataframes using `pd.concat()` method

In [25]:
df_concatenated = pd.concat(objs=[df_1985, df_1987, df_1997])

df_concatenated.shape

(3325, 12)

**Note the total number of rows in this dataframe equals to `1463+932+930 = 3325`**

In [ ]:
df_concatenated.head()

In [ ]:
# you can store the concatenated data inside your dataframe into a single Excel file
# You can mention the argument `index= false` for not storing row indices (0, 1,2,3,... in the Excel file.

df_concatenated.to_excel(excel_writer='temp.xlsx', index=False)

In [ ]:
# Let us verify
data = pd.read_excel(io='temp.xlsx')

data.shape

## 6. I/O with JSON Files

>**JSON**: JavaScript Object Notation is a text-based open standard file format that uses human-readable text consisting of attribute–value pairs and arrays. It is a data interchange format that is used to store and transfer the data via Internet, primarily between a web client and a server.

### a. Reading a Simple JSON File

In [ ]:
import sys
!{sys.executable} -m pip install SQLAlchemy psycopg2-binary

In [26]:
! cat datasets/simple.json

[{"name": "Kamal", "age": 12, "grade": "A"}, {"name": "Hashim", "age": 18, "grade": "B"}, {"name": "Salman", "age": 11, "grade": "A"}, {"name": "Mazhar", "age": 12, "grade": "C"}, {"name": "Eisha", "age": 13, "grade": "B"}, {"name": "Farhan", "age": 22, "grade": "C"}, {"name": "Mohsin", "age": 11, "grade": "A"}, {"name": "Bilal", "age": 19, "grade": "A"}, {"name": "Ishaan", "age": 10, "grade": "D"}, {"name": "Zalaid", "age": 9, "grade": "B"}]

In [27]:
# read the json file using read_json method of pandas library
df = pd.read_json('datasets/simple.json')
df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B

### b. Reading JSON File having each record in a separate line
- Some of the json files are written as records i.e each json line is a separate json object. For example:
```
{ 'name' : 'Ahsan', 'roll_no' : '100' } # line 1
{ 'name' : 'Ayesha' , 'roll_no' : '101' } # line 2
```

In [28]:
! cat datasets/simple_records.json

{"name": "Kamal", "age": 12, "grade": "A"}
{"name": "Hashim", "age": 18, "grade": "B"}
{"name": "Salman", "age": 11, "grade": "A"}
{"name": "Mazhar", "age": 12, "grade": "C"}
{"name": "Eisha", "age": 13, "grade": "B"}
{"name": "Farhan", "age": 22, "grade": "C"}
{"name": "Mohsin", "age": 11, "grade": "A"}
{"name": "Bilal", "age": 19, "grade": "A"}
{"name": "Ishaan", "age": 10, "grade": "D"}
{"name": "Zalaid", "age": 9, "grade": "B"}


In [5]:
# To read such file you need to pass `lines=True` to the `read_json()` method of dataframe
df = pd.read_json('datasets/simple_records.json',lines=True)

df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B

## 7. Writing Contents of Dataframe to a JSON File

In [ ]:
df.to_json('datasets/temp.json')

In [ ]:
df = pd.read_json('datasets/temp.json')
df